lock/common.go

```diff
@@ -0,0 +1,47 @@
+// Package lock provides implementations for cancellable and distributed locks.
+package lock
+
+import (
+	"time"
+
+	"gopkg.in/src-d/go-errors.v0"
+)
+
+var (
+	ErrCanceled      = errors.NewKind("context canceled")
+	ErrAlreadyClosed = errors.NewKind("already closed")
+)
+
+type SessionConfig struct {
+	TTL time.Duration
+}
+
+// Service is a locking service.
+type Service interface {
+	NewSession(*SessionConfig) (Session, error)
+	// Close closes the service and releases any resources held by it.
+	// If it is called more than once, ErrAlreadyClosed is returned.
+	Close() error
+}
+
+// Session is a locking session that can be reused to get multiple locks.
+// Multiple actors should use different sessions.
+type Session interface {
+	// NewService creates a NewService for the given id. Lockers returned for the
+	// same id on different sessions are mutually exclusive.
+	NewLocker(id string) Locker
+	// Close closes the session and releases any resources held by it.
+	// If it is called more than once, ErrAlreadyClosed is returned.
+	Close() error
+}
+
+// Locker is the interface to a lock.
+type Locker interface {
+	// Context can be used to cancel the operation (e.g. with a timeout).
+	// If it succeeds, it returns a cancel channel. Otherwise, it returns an
+	// error. The cancel channel is closed whenever the lock is not held anymore.
+	// Note that the lock might be lost before calling to Unlock.
+	Lock() (<-chan struct{}, error)
+	// Unlock releases the lock.
+	Unlock()
```

Comment:
Added an error and more documentation. Note that error is just informative (e.g. it might be logged) but no action needs to be taken on it, since locks should expire at some point.